### ISSUE TITLE
Metrics false report that a workflow is stuck in 'in_progress' status

### ISSUE THREAD

**Issue by ramir-savvy** [promhippie/github_exporter#296]
Hello,

**Summary:**
We get **false** metrics of workflows stuck in `in_progress` status. I believe related to GitHub's limitation that webhooks delivery order is [not guaranteed](https://docs.github.com/en/webhooks/testing-and-troubleshooting-webhooks/troubleshooting-webhooks#webhooks-deliveries-are-out-of-order):
> GitHub may deliver webhooks in a different order than the order in which the events took place. If you need to know when the event occurred relative to another event, you should use the timestamps that are included in the delivery payload. 

**Details:**
We have set an alert to detect workflows that are `in_progress` status for over 20 minutes: `github_workflow_duration_run_created_minutes{container="github-exporter", status="in_progress"} > 20` and we get those alerts once a day or so.
When looking at the allegedly `in_progress` workflow, we usually see that it was `skipped` or a short run time.

I don't have the `github-exporter` debug logs yet to prove it, but from my understanding of the code, the last webhook delivery event that arrives is the one that is stored in the database.
That could cause a problem if the webhook deliveries are not arriving by the order that they happened and according to GitHub, it is indeed a limitation and they [don't guarantee ordering](https://docs.github.com/en/webhooks/testing-and-troubleshooting-webhooks/troubleshooting-webhooks#webhooks-deliveries-are-out-of-order).

To try and prove it without the `github-exporter` logs (I will add those when I manage to capture it), I went over the related GitHub's App `Recent Deliveries` logs, and saw the following events related to the false reported long `in_progress` alert:

1. At `2023-12-20 04:51:17` GitHub delivered this `queued` workflow event`:
```
"action": "requested",
  "workflow_run": {
    "id": 7270372034,
...
    "run_number": 684,
    "event": "issue_comment",
    "status": "queued",
    "conclusion": null,
    "workflow_id": 45053147,
...
    "created_at": "2023-12-20T02:51:16Z",
    "updated_at": "2023-12-20T02:51:16Z",
...
	"run_attempt": 1,
...
	"run_started_at": "2023-12-20T02:51:16Z",
...
```
2. Then at `2023-12-20 04:51:20`  this `completed` event with `skipped` conclusion - this one should have been the last:
```
"action": "completed",
  "workflow_run": {
    "id": 7270372034,
...
    "run_number": 684,
    "event": "issue_comment",
    "status": "completed",
    "conclusion": "skipped",
    "workflow_id": 45053147,
...
    "created_at": "2023-12-20T02:51:16Z",
    "updated_at": "2023-12-20T02:51:19Z", <<<- mind this one later than the 'in_progress' event below
...
	"run_attempt": 1,
...
	"run_started_at": "2023-12-20T02:51:16Z",
...
```
3. And at `2023-12-20 04:51:21` the `in_progress` event - this one should have been the second:
```
"action": "in_progress",
  "workflow_run": {
    "id": 7270372034,
...
    "run_number": 684,
    "event": "issue_comment",
    "status": "in_progress",
    "conclusion": null,
    "workflow_id": 45053147,
...
    "created_at": "2023-12-20T02:51:16Z",
    "updated_at": "2023-12-20T02:51:17Z",  <<<- mind that this one is earlier than the 'completed' event above
...
	"run_attempt": 1,
...
	"run_started_at": "2023-12-20T02:51:16Z",
...
```

I would like to suggest a fix, that when a webhook event arrives, it will be dropped if its `updated_at` time is earlier than the one that is stored in the DB.
Something like adding the following code block before calling `updateWorkflowRunQuery` [here](https://github.com/promhippie/github_exporter/blob/655fa65cdb65d6271ac0e5680adde14cc79bb5e8/pkg/store/generic.go#L65):
```
if existing.UpdatedAt > record.UpdatedAt {
	return nil
}
```

Looking forward to your input.

Thanks

**Comment by tboerger** [comment]
A good catch and your suggestion sounds like the best solution for the issue. Thanks for the detailed investigation :)

**Comment by ramir-savvy** [comment]
@tboerger, not sure if you are waiting for me to submit a PR on that or if you are planning to fix it soon? Thanks again

**Comment by tboerger** [comment]
Currently I'm more or less on leave for christmas. Feel free to create a PR including a changelog entry :)

**Comment by ramir-savvy** [comment]
Got it. Will submit one tomorrow. Merry Christmas!

**Comment by ramir-savvy** [comment]
@tboerger, I am running with `latest` image that should include this recent fix, but I just ran into the same situation as the one I reported in this ticker: the `complete` event arrived before the `in_progress` event, but for some reason, the `in_progress` event wasn't rejected. I will try to see if I can debug it locally and add debugs to the logs

**Comment by tboerger** [comment]
The release including the fix is in progress, there will be v3.1.1 in a few minutes.
